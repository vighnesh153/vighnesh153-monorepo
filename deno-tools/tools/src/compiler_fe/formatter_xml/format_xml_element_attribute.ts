import type { XmlElementAttribute } from "@/compiler_fe/parser_xml/mod.ts";

export function formatXmlElementAttribute(
  { namespaces, value }: XmlElementAttribute,
): string {
  const combinedNamespace = namespaces.map((namespace) =>
    namespace.tokenLiteral
  ).join(":");
  return `${combinedNamespace}="${value.tokenLiteral}"`;
}
