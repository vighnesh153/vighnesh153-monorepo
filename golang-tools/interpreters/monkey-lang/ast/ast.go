package ast

import "vighnesh153/interpreters/monkey-lang/token"

type Node interface {
	TokenLiteral() string
}

type Statement interface {
	Node
	statementNode()
}

type Expression interface {
	Node
	expressionNode()
}

type Program struct {
	Statements []Statement
}

func (p *Program) TokenLiteral() string {
	if len(p.Statements) > 0 {
		return p.Statements[0].TokenLiteral()
	} else {
		return ""
	}
}

type LetStatement struct {
	Token token.Token // the token.LET token
	Name  *Identifier
	Value Expression
}

func (ls *LetStatement) statementNode()       {}
func (ls *LetStatement) TokenLiteral() string { return ls.Token.Literal }

type Identifier struct {
	Token token.Token // token.IDENTIFIER token
	Value string
}

func (i *Identifier) expressionNode()      {}
func (i *Identifier) TokenLiteral() string { return i.Token.Literal }

type ReturnStatement struct {
	Token       token.Token // the token.RETURN token
	ReturnValue Expression
}

func (ls *ReturnStatement) statementNode()       {}
func (ls *ReturnStatement) TokenLiteral() string { return ls.Token.Literal }
