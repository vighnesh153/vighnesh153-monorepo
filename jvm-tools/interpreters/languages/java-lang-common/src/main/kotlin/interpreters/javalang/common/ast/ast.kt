package interpreters.javalang.common.ast

import interpreters.javalang.common.tokens.Token
import interpreters.javalang.common.tokens.TokenType
import kotlin.text.StringBuilder

abstract class LanguageNode {
    abstract fun tokenLiteral(): String

    abstract override fun toString(): String
}

abstract class StatementNode : LanguageNode()
abstract class ExpressionNode : LanguageNode()

class ProgramNode : LanguageNode() {
    private val statements = mutableListOf<StatementNode>()

    fun getStatements(): List<StatementNode> = statements

    fun addStatement(statement: StatementNode) {
        this.statements.add(statement)
    }

    override fun toString(): String = StringBuilder().apply {
        statements.forEach {
            append(it.toString())
            append("\n")
        }
    }.toString()

    override fun tokenLiteral(): String = "<PROGRAM>"
}

class PackageStatement(
    val token: Token,  // The TokenType.PACKAGE_KEYWORD token
    val dotSeparatedIdentifiers: List<Token>,
) : StatementNode() {

    override fun tokenLiteral(): String {
        return token.tokenLiteral
    }

    override fun toString(): String {
        return "${token.tokenLiteral} ${dotSeparatedIdentifiers.joinToString { "." }};"
    }
}

class ImportStatement(
    val token: Token,  // The TokenType.IMPORT_KEYWORD token
    val dotSeparatedIdentifiers: List<Token>,
    val isStaticImport: Boolean,
) : StatementNode() {

    override fun tokenLiteral(): String {
        return token.tokenLiteral
    }

    override fun toString(): String {
        if (isStaticImport) {
            return "${token.tokenLiteral} static ${dotSeparatedIdentifiers.joinToString { "." }};"
        }
        return "${token.tokenLiteral} ${dotSeparatedIdentifiers.joinToString { "." }};"
    }
}

class ExpressionStatement(
    val token: Token,
    val expression: ExpressionNode?,
): StatementNode() {
    override fun tokenLiteral(): String {
        return expression?.tokenLiteral() ?: token.tokenLiteral
    }

    override fun toString(): String {
        return expression?.toString() ?: token.toString()
    }
}

class IntegerLiteral(
    val token: Token,
    val value: Int,
) : ExpressionNode() {
    override fun tokenLiteral(): String {
        return token.tokenLiteral
    }

    override fun toString(): String {
        return token.tokenLiteral
    }
}

class FloatLiteral(
    val token: Token,
    val value: Float,
) : ExpressionNode() {
    override fun tokenLiteral(): String {
        return token.tokenLiteral
    }

    override fun toString(): String {
        return token.tokenLiteral
    }
}

class LongLiteral(
    val token: Token,
    val value: Long,
) : ExpressionNode() {
    override fun tokenLiteral(): String {
        return token.tokenLiteral
    }

    override fun toString(): String {
        return token.tokenLiteral
    }
}

class DoubleLiteral(
    val token: Token,
    val value: Double,
) : ExpressionNode() {
    override fun tokenLiteral(): String {
        return token.tokenLiteral
    }

    override fun toString(): String {
        return token.tokenLiteral
    }
}

class StringLiteral(
    val token: Token,
    val value: String,
) : ExpressionNode() {
    override fun tokenLiteral(): String {
        return value
    }

    override fun toString(): String {
        return token.tokenLiteral
    }
}

class CharacterLiteral(
    val token: Token,
    val value: Char,
) : ExpressionNode() {
    override fun tokenLiteral(): String {
        return token.tokenLiteral
    }

    override fun toString(): String {
        return "$value"
    }
}
