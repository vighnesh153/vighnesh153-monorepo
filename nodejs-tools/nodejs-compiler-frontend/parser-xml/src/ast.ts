/* eslint-disable @typescript-eslint/no-use-before-define */

import { Token } from '@vighnesh153/lexer-xml';

export interface XmlExpression {
  toString(indentation: number): string;
}

export class XmlElementProperty {
  constructor(
    public readonly colonSeparatedKeys: readonly Token[],
    public readonly value: Token
  ) {}

  toString(): string {
    return `${this.colonSeparatedKeys.map((key) => key.tokenLiteral).join(':')}="${this.value.tokenLiteral}"`;
  }
}

export class XmlProgram implements XmlExpression {
  #statements: XmlExpression[] = [];

  get statements(): readonly XmlExpression[] {
    return [...this.#statements];
  }

  addStatement(statement: XmlExpression): void {
    this.#statements.push(statement);
  }

  toString(indentation: number): string {
    return this.#statements.map((statement) => statement.toString(indentation)).join('\n');
  }
}

// <?xml version="1.0"?>
export class XmlPrologNode implements XmlExpression {
  #properties: XmlElementProperty[] = [];

  get properties(): readonly XmlElementProperty[] {
    return [...this.#properties];
  }

  addProperty(property: XmlElementProperty): void {
    this.#properties.push(property);
  }

  toString(indentation: number = 0): string {
    const stringBuilder: string[] = [];

    stringBuilder.push(`${buildIndentationSpace(indentation)}<?xml`);
    for (const property of this.properties) {
      stringBuilder.push(property.toString());
    }

    return stringBuilder.join(' ') + `?>`;
  }
}

// <manifest />
export class XmlTagNode implements XmlExpression {
  #tagIdentifier: Token;

  #properties: XmlElementProperty[] = [];
  #children: XmlExpression[] = [];

  get tagIdentifier(): Token {
    return this.#tagIdentifier;
  }

  get properties(): readonly XmlElementProperty[] {
    return [...this.#properties];
  }

  get children(): readonly XmlExpression[] {
    return [...this.#children];
  }

  constructor(tagIdentifier: Token) {
    this.#tagIdentifier = tagIdentifier;
  }

  addProperty(property: XmlElementProperty): void {
    this.#properties.push(property);
  }

  addChild(statement: XmlExpression): void {
    this.#children.push(statement);
  }

  toString(indentation: number = 0): string {
    const { tagIdentifier, properties, children } = this;

    const stringBuilder: string[] = [];

    stringBuilder.push(`${buildIndentationSpace(indentation)}<${tagIdentifier.tokenLiteral}`);
    if (properties.length === 1) {
      const serializedProps = properties.map((property) => property.toString()).join(' ');
      stringBuilder[stringBuilder.length - 1] += ' ' + serializedProps;
    } else if (properties.length > 1) {
      properties.forEach((property) => {
        stringBuilder.push(`${buildIndentationSpace(indentation + 1)}${property.toString()}`);
      });
    }

    if (children.length === 0) {
      stringBuilder[stringBuilder.length - 1] = stringBuilder.at(-1) + ' />';
    } else {
      stringBuilder[stringBuilder.length - 1] = stringBuilder.at(-1) + '>';
    }

    for (const child of children) {
      stringBuilder.push(child.toString(indentation + 1));
    }

    if (children.length > 0) {
      stringBuilder.push(`${buildIndentationSpace(indentation)}</${tagIdentifier.tokenLiteral}>`);
    }

    return stringBuilder.join('\n');
  }
}

export class XmlCommentNode implements XmlExpression {
  constructor(public readonly comment: Readonly<Token>) {}

  toString(indentation: number): string {
    return `${buildIndentationSpace(indentation)}<!-- ${this.comment.tokenLiteral.trim()} -->`;
  }
}

function buildIndentationSpace(indentation: number): string {
  return Array.from({ length: indentation * 4 })
    .map(() => ' ')
    .join('');
}
