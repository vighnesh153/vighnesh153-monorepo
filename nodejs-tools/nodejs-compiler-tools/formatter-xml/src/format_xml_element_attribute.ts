import { XmlElementAttribute } from '@vighnesh153/parser-xml';

export function formatXmlElementAttribute({ namespaces, value }: XmlElementAttribute): string {
  const combinedNamespace = namespaces.map((namespace) => namespace.tokenLiteral).join(':');
  return `${combinedNamespace}="${value.tokenLiteral}"`;
}
