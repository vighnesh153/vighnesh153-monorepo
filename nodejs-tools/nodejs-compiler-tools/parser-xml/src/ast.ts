/* eslint-disable @typescript-eslint/no-use-before-define */

import { Token } from '@vighnesh153/lexer-xml';

export type AstNodeType = 'XML_PROGRAM' | 'XML_PROLOG_NODE' | 'XML_TAG_NODE' | 'XML_COMMENT_NODE';

export interface XmlExpression {
  readonly astNodeType: AstNodeType;

  toString(indentation: number): string;
}

export class XmlElementProperty {
  constructor(
    public readonly colonSeparatedKeys: readonly Token[],
    public readonly value: Token
  ) {}

  toString(): string {
    return `${this.colonSeparatedKeys.map((key) => key.tokenLiteral).join(':')}="${this.value.tokenLiteral}"`;
  }
}

export class XmlProgram implements XmlExpression {
  readonly astNodeType: AstNodeType = 'XML_PROGRAM';

  #statements: XmlExpression[] = [];

  get statements(): readonly XmlExpression[] {
    return [...this.#statements];
  }

  addStatement(statement: XmlExpression): void {
    this.#statements.push(statement);
  }

  toString(indentation: number): string {
    return this.#statements.map((statement) => statement.toString(indentation)).join('\n');
  }
}

// <?xml version="1.0"?>
export class XmlPrologNode implements XmlExpression {
  readonly astNodeType: AstNodeType = 'XML_PROLOG_NODE';

  #properties: XmlElementProperty[] = [];

  get properties(): readonly XmlElementProperty[] {
    return [...this.#properties];
  }

  addProperty(property: XmlElementProperty): void {
    this.#properties.push(property);
  }

  toString(indentation: number = 0): string {
    const stringBuilder: string[] = [];

    stringBuilder.push(`${buildIndentationSpace(indentation)}<?xml`);
    for (const property of this.properties) {
      stringBuilder.push(property.toString());
    }

    return stringBuilder.join(' ') + `?>`;
  }
}

// <manifest />
export class XmlTagNode implements XmlExpression {
  readonly astNodeType: AstNodeType = 'XML_TAG_NODE';

  #tagIdentifier: Token;

  #properties: XmlElementProperty[] = [];
  #children: XmlExpression[] = [];

  get tagIdentifier(): Token {
    return this.#tagIdentifier;
  }

  get properties(): readonly XmlElementProperty[] {
    return [...this.#properties];
  }

  get children(): readonly XmlExpression[] {
    return [...this.#children];
  }

  constructor(tagIdentifier: Token) {
    this.#tagIdentifier = tagIdentifier;
  }

  addProperty(property: XmlElementProperty): void {
    this.#properties.push(property);
  }

  addChild(statement: XmlExpression): void {
    this.#children.push(statement);
  }

  toString(indentation: number = 0): string {
    const { tagIdentifier, properties, children } = this;

    const stringBuilder: string[] = [];

    stringBuilder.push(`${buildIndentationSpace(indentation)}<${tagIdentifier.tokenLiteral}`);
    if (properties.length === 1) {
      const serializedProps = properties.map((property) => property.toString()).join(' ');
      stringBuilder[stringBuilder.length - 1] += ' ' + serializedProps;
    } else if (properties.length > 1) {
      properties.forEach((property) => {
        stringBuilder.push(`${buildIndentationSpace(indentation + 1)}${property.toString()}`);
      });
    }

    if (children.length === 0) {
      stringBuilder[stringBuilder.length - 1] = stringBuilder.at(-1) + ' />';
    } else {
      stringBuilder[stringBuilder.length - 1] = stringBuilder.at(-1) + '>';
    }

    for (const child of children) {
      stringBuilder.push(child.toString(indentation + 1));
    }

    if (children.length > 0) {
      stringBuilder.push(`${buildIndentationSpace(indentation)}</${tagIdentifier.tokenLiteral}>`);
    }

    return stringBuilder.join('\n');
  }
}

export class XmlCommentNode implements XmlExpression {
  readonly astNodeType: AstNodeType = 'XML_COMMENT_NODE';

  constructor(public readonly comment: Readonly<Token>) {}

  toString(indentation: number): string {
    return `${buildIndentationSpace(indentation)}<!-- ${this.comment.tokenLiteral.trim()} -->`;
  }
}

function buildIndentationSpace(indentation: number): string {
  return Array.from({ length: indentation * 4 })
    .map(() => ' ')
    .join('');
}
