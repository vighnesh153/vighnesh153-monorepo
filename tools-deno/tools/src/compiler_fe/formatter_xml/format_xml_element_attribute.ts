import type { XmlElementAttribute } from "@/compiler_fe/parser_xml/mod.ts";

export function formatXmlElementAttribute(
  attribute: XmlElementAttribute,
): string {
  return `${attribute.attributeName.tokenLiteral}="${attribute.value.tokenLiteral}"`;
}
