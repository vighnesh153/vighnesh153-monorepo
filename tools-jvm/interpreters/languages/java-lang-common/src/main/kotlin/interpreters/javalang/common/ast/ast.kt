package interpreters.javalang.common.ast

import interpreters.javalang.common.tokens.Token
import kotlin.text.StringBuilder

abstract class LanguageNode {
    abstract fun tokenLiteral(): String

    abstract override fun toString(): String
}

abstract class StatementNode : LanguageNode()
abstract class ExpressionNode : LanguageNode()

class ProgramNode : LanguageNode() {
    private val statements = mutableListOf<StatementNode>()

    fun getStatements(): List<StatementNode> = statements

    fun addStatement(statement: StatementNode) {
        this.statements.add(statement)
    }

    override fun toString(): String = StringBuilder().apply {
        statements.forEach {
            append(it.toString())
            append("\n")
        }
    }.toString()

    override fun tokenLiteral(): String = "<PROGRAM>"
}

class PackageStatement(
    val token: Token,  // The TokenType.PACKAGE_KEYWORD token
    val dotSeparatedIdentifiers: List<Token>,
) : StatementNode() {

    override fun tokenLiteral(): String {
        return token.tokenLiteral
    }

    override fun toString(): String {
        return "${token.tokenLiteral} ${dotSeparatedIdentifiers.joinToString { "." }};"
    }
}

class ImportStatement(
    val token: Token,  // The TokenType.IMPORT_KEYWORD token
    val dotSeparatedIdentifiers: List<Token>,
    val isStaticImport: Boolean,
) : StatementNode() {

    override fun tokenLiteral(): String {
        return token.tokenLiteral
    }

    override fun toString(): String {
        if (isStaticImport) {
            return "${token.tokenLiteral} static ${dotSeparatedIdentifiers.joinToString { "." }};"
        }
        return "${token.tokenLiteral} ${dotSeparatedIdentifiers.joinToString { "." }};"
    }
}

class ExpressionStatement(
    val token: Token,
    val expression: ExpressionNode?,
) : StatementNode() {
    override fun tokenLiteral(): String {
        return expression?.tokenLiteral() ?: token.tokenLiteral
    }

    override fun toString(): String {
        return expression?.toString() ?: token.toString()
    }
}

class PrefixExpression(
    val operatorToken: Token,
    val right: ExpressionNode,
) : ExpressionNode() {
    override fun tokenLiteral(): String {
        return "${operatorToken.tokenLiteral}${right.tokenLiteral()}"
    }

    override fun toString(): String {
        return "${operatorToken.tokenLiteral}$right"
    }
}

class PostfixExpression(
    val operatorToken: Token,
    val left: ExpressionNode,
) : ExpressionNode() {
    override fun tokenLiteral(): String {
        return "${left.tokenLiteral()}${operatorToken.tokenLiteral}"
    }

    override fun toString(): String {
        return "$left${operatorToken.tokenLiteral}"
    }
}

class InfixExpression(
    val operatorToken: Token,
    val left: ExpressionNode,
    val right: ExpressionNode,
) : ExpressionNode() {
    override fun tokenLiteral(): String {
        return "(${left.tokenLiteral()} ${operatorToken.tokenLiteral} ${right.tokenLiteral()})"
    }

    override fun toString(): String {
        return "($left ${operatorToken.tokenLiteral} $right)"
    }
}

