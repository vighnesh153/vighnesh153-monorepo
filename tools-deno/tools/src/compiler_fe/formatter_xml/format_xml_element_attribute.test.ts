import { assertSnapshot } from "@std/testing/snapshot";
import type { Token } from "@/compiler_fe/lexer_core/mod.ts";
import { XmlTokenType } from "@/compiler_fe/lexer_xml/mod.ts";
import { XmlElementAttribute } from "@/compiler_fe/parser_xml/ast.ts";
import { formatXmlElementAttribute } from "./format_xml_element_attribute.ts";

function createToken(
  tokenLiteral: string,
  tokenType: XmlTokenType,
): Token<XmlTokenType> {
  return {
    lineNumber: 1,
    columnNumber: 1,
    tokenLiteral,
    tokenType,
  } satisfies Token<XmlTokenType>;
}

Deno.test("formatXmlElementAttribute should format attribute with single namespace", async (t) => {
  await assertSnapshot(
    t,
    formatXmlElementAttribute(
      new XmlElementAttribute(
        createToken("pokemon", XmlTokenType.Identifier),
        createToken("pikachu", XmlTokenType.StringLiteral),
      ),
    ),
  );
});

Deno.test("formatXmlElementAttribute should format attribute with multiple namespaces", async (t) => {
  await assertSnapshot(
    t,
    formatXmlElementAttribute(
      new XmlElementAttribute(
        createToken("pokemon.electric.yellow", XmlTokenType.Identifier),
        createToken("pikachu", XmlTokenType.StringLiteral),
      ),
    ),
  );
});
