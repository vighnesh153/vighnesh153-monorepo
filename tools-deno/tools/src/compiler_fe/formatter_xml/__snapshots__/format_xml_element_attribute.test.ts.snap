export const snapshot = {};

snapshot[`formatXmlElementAttribute should format attribute with single namespace 1`] = `'pokemon="pikachu"'`;

snapshot[`formatXmlElementAttribute should format attribute with multiple namespaces 1`] = `'pokemon.electric.yellow="pikachu"'`;
